### ISSUE TITLE
Styles cannot be dequeued on the [affiliate_login] shortcode

### ISSUE THREAD

**Issue by sumobi** [AffiliateWP/AffiliateWP#1670]
The standalone login form which uses the `[affiliate_login]` shortcode outputs its CSS file directly inside the `affiliate_login()` method.  This isn't ideal for 2 reasons:
1. It loads the `forms.css` file just before the closing `</body>` of the page. CSS files should generally be loaded in the `<head>` 
2. The `forms.css` file cannot be dequeued so a theme can't use its own CSS styling for the form.

Like we did for https://github.com/AffiliateWP/AffiliateWP/issues/1026 I propose we remove the `wp_enqueue_style` function call and load the stylesheet inside our `affwp_frontend_scripts_and_styles()` function since the CSS has already been registered. 

This will allow a theme to dequeue the CSS file, as well as the forms.css file being correctly loaded in the `<head>`.

**Comment by DrewAPicture** [comment]
See the PR on #1672, which (indirectly) addresses this.

**Comment by DrewAPicture** [comment]
After some discussion with @sumobi the other day in Slack, our best option to make scripts or styles dequeueable might be something along these lines:
```php
function affwp_enqueue_style( $handle, $context = '' ) {
    /**
     * Filter whether to enqueue the stylesheet.
     * 
     * The dynamic portion of the hook name, `$handle` refers to the stylesheet handle.
     * 
     * @since 1.9.4
     * 
     * @param bool   $enqueue Whether to enqueue the stylesheet. Default true.
     * @param string $context Context under which to enqueue the stylesheet.
     */
    if ( true === apply_filters( "affwp_enqueue_style_{$handle}", true, $context ) ) {
        wp_enqueue_style( $handle );
    }
}
```

 ```php
add_filter( 'affwp_enqueue_style_affwp-forms', function( $enqueue, $context ) {
    if ( 'login' === $context' ) {
        $enqueue = false;
    }
    return $enqueue;
} );
```

**Comment by pippinsplugins** [comment]
I like that proposal.

**Comment by DrewAPicture** [comment]
OK. I'll work on a PR.

**Comment by DrewAPicture** [comment]
I've added a secondary PR with the `affwp_enqueue_*()` approach. It avoids using `has_shortcode()` in a couple of places and simultaneously introduces the ability to selectively disable enqueuing within individual shortcode context. Please test :-)

**Comment by pippinsplugins** [comment]
I really like this approach.

**Comment by sumobi** [comment]
@DrewAPicture awesome, will check it out

**Comment by sumobi** [comment]
Oh yeah this is slick, love it + works great

**Comment by sumobi** [comment]
Although it's now easy as pie to dequeue the stylesheet via the filters, there's still point 1 to address above. 

This is also mentioned here: https://github.com/AffiliateWP/AffiliateWP/issues/1672#issuecomment-255248538

Imagine a customer/theme developer wanting to make a small change to the CSS styling. Adding this to their stylesheet would not work:

```
.affwp-form input {
	border: 1px solid red;
}
```

Sure you could be more specific with the selectors and this would work:

```
#affwp-login-form.affwp-form input {
	border: 1px solid red;
}
```

The point though is that it shouldn't require any additional selectors. If the theme stylesheet is loaded after ours you simply need to use the exact same CSS selector to override.

What really makes my eyes bleed though is how you see the un-styled forms before the page finishes loading. CSS should be loaded in the `<head>`, _before_ the HTML markup.

Since the forms.css is loaded right at the end of the document, the unstyled form appears for a brief moment until the CSS kicks in. Refresh your browser a few times to see this, or throttle your connection.

Any ideas?

**Comment by sumobi** [comment]
Example

https://cl.ly/0l1L14390Y35

**Comment by DrewAPicture** [comment]
In terms of third-parties customizing styles, they can actually add it in-line to the registered stylesheet it using `wp_add_inline_style()`, e.g.
```php
wp_add_iniline_style( 'affwp-forms', '.affwp-form input { border: 1px solid red; }' );
```

**Comment by sumobi** [comment]
Good to know! Some however would prefer to put styles directly in their stylesheet, rather than writing a function.

**Comment by DrewAPicture** [comment]
Right. So we have to decide. We can either:

1. Load it all the time in the header
2. Use a regex to search the entire content for a given shortcode and then enqueue it in the header
3. Selectively enqueue it on the spot

On number 1, this would honestly be my choice, when it's minified, it's another HTTP request but it's very small and doesn't add much to the page.

On number 2 (the current solution), we could use `has_shortcode()` to run a costly regex against the post content and then enqueue the stylesheet if it's found. Note that the regex runs every time, whether the shortcode is in there or not. It's basically the equivalent of #1 with a greater performance hit. Every time the page loads ;)

On number 3, it's the most efficient but the least user friendly because you have to jump through hoops to override it from something like a theme stylesheet. Though it's worth noting that the theme stylesheet also might be enqueued _after_ forms.css is depending entirely on how it's set up.

**Comment by DrewAPicture** [comment]
Looking for feedback from @AffiliateWP/core on the attached PR as well as [this comment](https://github.com/AffiliateWP/AffiliateWP/issues/1670#issuecomment-256221864).

**Comment by pippinsplugins** [comment]
I vote for number 1.

**Comment by DrewAPicture** [comment]
Adjusted PR #1748 to always enqueue the `affwp-forms` stylesheet via `wp_enqueue_scripts`. It still uses `affwp_enqueue_style()` so it's still filterable.

**Comment by DrewAPicture** [comment]
I believe #1748 just needs a look-see and a test to ensure the forms stylesheet is always enqueued but can be dequeued selectively. Closed #1672 in favor of this issue since it also addresses the duplicate enqueues.

**Comment by pippinsplugins** [comment]
@DrewAPicture could you resolve the merge conflict please?

**Comment by DrewAPicture** [comment]
@pippinsplugins I've cleared the merge conflict and updated from master. Tests are passing.

**Comment by sumobi** [comment]
Happy with this

**Comment by DrewAPicture** [comment]
Fixed the admin menu dashicon in f8407be.
